### PR TITLE
ScalarFunction objective for MOI and JuMP

### PR DESCRIPTION
**Pull request by Azzaare**

### CODE REVIEW

File: src/MOI_wrapper/objectives.jl [hunk 1]
```diff
@@ -1,14 +1,40 @@
-# TODO: MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
-MOI.supports(::Optimizer, ::MOI.ObjectiveFunction{ScalarFunction{F}}) where {F <: Function} = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
+MOI.get(model::Optimizer, ::MOI.ObjectiveSense) = Min
```
Comment (odow) [inline src/MOI_wrapper/objectives.jl:2]: Did you define `Min`? It should be `MOI.MIN_SENSE`.

File: src/MOI_wrapper/objectives.jl [hunk 2]
```diff
@@ -1,14 +1,40 @@
...
+
+    ScalarFunction(f, X::Union{Nothing, VOV} = nothing) = (@warn X; new{typeof(f), typeof(X)}(f, X))
+end
+
+# external constructors
+
+function ScalarFunction(f, X::A) where {A <: AbstractArray{VariableRef}}
+    return ScalarFunction(f, VOV(vec(map(index, X))))
+end
+ScalarFunction(f, x::VariableRef) = ScalarFunction(f, [x])
```
Comment (odow) [inline src/MOI_wrapper/objectives.jl:22]: These are a nice solution to supporting JuMP variables.

File: src/MOI_wrapper/objectives.jl [hunk 3]
```diff
@@ -1,14 +1,40 @@
...
-# end
+# supports
+MOI.supports(::Optimizer, ::OF{ScalarFunction{F, V}}) where {F <: Function, V <: Union{Nothing, SVF,VOV}} = true
 
-function MOI.set(optimizer::Optimizer, ::MOI.ObjectiveFunction, func::ScalarFunction)
-    objective!(optimizer, func.f)
+# set
+function MOI.set(optimizer::Optimizer, ::OF, func::ScalarFunction{F, V}
+) where {F <: Function, V <: Union{Nothing, VOV}}
+    objective_function = if isnothing(func.X) # for direct use of MOI
```
Comment (odow) [inline src/MOI_wrapper/objectives.jl:34]: Best practice is to write two functions 
```Julia
function MOI.set(optimizer::Optimizer, ::OF{F}, func::F} where {F<:ScalarFunction{<:Function,Nothing}}
# and
function MOI.set(optimizer::Optimizer, ::OF{F}, func::F} where {F<:ScalarFunction{<:Function,VOV}}
```

**Review by odow — COMMENTED** [review]
From the error you mentioned on Gitter, you need to implement:
https://github.com/jump-dev/MathOptInterface.jl/blob/f71ed8314da336a90f5713e2406ca4007c8c02e7/src/Utilities/functions.jl#L55-L67
Here's an example:
https://github.com/jump-dev/MathOptInterface.jl/blob/f71ed8314da336a90f5713e2406ca4007c8c02e7/src/Utilities/functions.jl#L109-L111

File: src/MOI_wrapper/objectives.jl (hunk 1 above)
Comment (Azzaare) [reply src/MOI_wrapper/objectives.jl:2]: Good catch! Thanks

File: src/MOI_wrapper/objectives.jl (hunk 3 above)
Comment (Azzaare) [reply src/MOI_wrapper/objectives.jl:34]: You're definitely right! This code is a mess ... I need to clean things up

**Pull request merged**